internal/app/db.go

```diff
@@ -25,30 +26,218 @@ type Column struct {
+
+	// configs
+	windowInterval time.Duration
+
+	// lock
+	mu sync.Mutex
+
+	// control attributes
+	close chan struct{}
 }
 
 // NewDBManager creates a new DBManager.
-func NewDBManager(dbDir, table string, cols []Column, winSize time.Duration, uploader *VaultsUploader) *DBManager {
+func NewDBManager(
+	dbDir, table string, cols []Column, windowInterval time.Duration, uploader *VaultsUploader,
+) *DBManager {
 	return &DBManager{
-		dbDir:    dbDir,
-		table:    table,
-		cols:     cols,
-		winSize:  winSize,
-		uploader: uploader,
+		dbDir:          dbDir,
+		table:          table,
+		cols:           cols,
+		windowInterval: windowInterval,
+		uploader:       uploader,
+	}
+}
+
+// NewDB creates a new duckdb database at the <ts>.db path.
+func (dbm *DBManager) NewDB(ctx context.Context) error {
+	now := time.Now()
+	dbm.dbFname = fmt.Sprintf("%d.db", now.UnixNano())
+	dbPath := path.Join(dbm.dbDir, dbm.dbFname)
+	db, err := sql.Open("duckdb", dbPath)
+	if err != nil {
+		return fmt.Errorf("cannot open db: %s", err)
+	}
+
+	slog.Info("created new db", "at", dbPath)
+	dbm.db = db
+
+	if err := dbm.setup(ctx); err != nil {
+		return fmt.Errorf("cannot setup db: %s", err)
 	}
+
+	ticker := time.NewTicker(dbm.windowInterval)
+	dbm.close = make(chan struct{})
+
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				dbm.mu.Lock()
+				slog.Info("window interval passed")
+				if err := dbm.replace(ctx); err != nil {
+					slog.Error("replacing current db before replaying further txs", "error", err)
+				}
+				dbm.mu.Unlock()
+			case <-dbm.close:
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Replay replays a WAL record onto the current db.
+// If the window has passed, it replaces the current db with
+// a new one. The current db is exported and uploaded before
+// new db is ready to be used.
+func (dbm *DBManager) Replay(ctx context.Context, tx *pgrepl.Tx) error {
+	dbm.mu.Lock()
+	defer dbm.mu.Unlock()
+
+	query, err := dbm.queryFromWAL(tx)
+	if err != nil {
+		return err
+	}
+
+	slog.Info("replaying", "query", query)
+	_, err = dbm.db.ExecContext(ctx, query)
+	if err != nil {
+		return fmt.Errorf("cannot replay WAL record: %v", err)
+	}
+
+	return nil
+}
+
+// Export exports the current db to a parquet file at the given path.
+func (dbm *DBManager) Export(ctx context.Context, exportPath string) (bool, error) {
+	var err error
+	db := dbm.db
+	// db is nil before replication starts.
+	// In that case, we open all existing db files
+	// and upload them.
+	if db == nil {
+		// convert the export path to a db path:
+		// <ts>.db.parquet -> <ts>.db
+		dbPath := strings.ReplaceAll(exportPath, ".parquet", "")
+		db, err = sql.Open("duckdb", dbPath)
+		if err != nil {
+			return true, err
+		}
+		defer func() {
+			if err := db.Close(); err != nil {
+				fmt.Printf("cannot close db %v \n", err)
+			}
+		}()
+		slog.Info("backing up db", "at", exportPath)
+	} else {
+		slog.Info("backing up current db")
+	}
+	var n int
```

Comment:
this query helps us figure out if the database is empty or not, so we don't upload an empty parquet file